PagingWithNetworkSample/benchmark/src/androidTest/java/com/example/benchmark/BenchmarkActivity.kt

```diff
@@ -0,0 +1,60 @@
+package com.example.benchmark
+
+import android.os.Bundle
+import android.util.Log
+import androidx.appcompat.app.AppCompatActivity
+import androidx.paging.PageKeyedDataSource
+import androidx.paging.PagedList
+import com.android.example.paging.pagingwithnetwork.GlideApp
+import com.android.example.paging.pagingwithnetwork.reddit.repository.NetworkState
+import com.android.example.paging.pagingwithnetwork.reddit.ui.PostsAdapter
+import com.android.example.paging.pagingwithnetwork.reddit.vo.RedditPost
+import kotlinx.android.synthetic.main.activity_benchmark.*
+import java.util.concurrent.Executors
+
+class BenchmarkActivity : AppCompatActivity() {
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        setContentView(R.layout.activity_benchmark)
+
+        val glide = GlideApp.with(this)
+        val adapter = PostsAdapter(glide) {}
+        list.adapter = adapter
+
+        val config = PagedList.Config.Builder()
+                .setInitialLoadSizeHint(5)
+                .setPageSize(5)
+                .build()
+
+        val pagedStrings: PagedList<RedditPost> = PagedList.Builder<Int, RedditPost>(MockDataSource(), config)
+                .setInitialKey(0)
+                .setFetchExecutor(Executors.newSingleThreadExecutor())
```

Comment:
Hmm, this is a bit of a problem - posting back and forth wont work since the benchmark is run as a single task in the UI thread - we're never letting the messages notifying of new pages being available to run.

Can you make both of these instant executors?

I think in general posting won't work, but I'm worried that will make this unrealistic. It will, for example, get the list item generation back onto the UI thread...

One alternative is to use a custom executor for both, and flush all tasks at the end of the benchmark loop, while timing is disabled. Does that seem more elegant?

Reply:
:+1:

Reply:
Actually - it turns out running the fetch executor on the main UI thread causes this error during benchmark:

java.lang.IllegalStateException: Cannot call this method while RecyclerView is computing a layout or scrolling androidx.recyclerview.widget.RecyclerView{25519d6 VFED.V... ......ID 0,0-1080,1536 #7f080071 app:id/list},

Reply:
Try the custom executor approach then (sort of like an Rx TestScheduler).